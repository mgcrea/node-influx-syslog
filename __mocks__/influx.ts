// @NOTE https://jestjs.io/docs/en/manual-mocks.html

const d = (obj: any) => console.dir(obj, {depth: 10, colors: true});

const influx = jest.genMockFromModule('influx');

const __mockedWritePoints = jest.fn().mockImplementation(() => {
  d('writePoints');
});

export class InfluxDB {
  addSchema() {}
  async writePoints() {
    __mockedWritePoints();
  }
}

influx.InfluxDB = InfluxDB;
influx.__mockedWritePoints = __mockedWritePoints;

module.exports = influx;
