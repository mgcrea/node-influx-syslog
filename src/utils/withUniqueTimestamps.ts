import {IPoint} from 'influx';
import padStart from 'lodash.padstart';

const withUniqueTimestamps = (points: IPoint[]): IPoint[] => {
  const initialValue: {
    uniquePoints: IPoint[];
    lastTimestamp: number;
    currentNanoSecond: number;
  } = {uniquePoints: [], lastTimestamp: 0, currentNanoSecond: 0};

  const {uniquePoints} = points.reduce((soFar, value) => {
    // Only process fields with timestamps
    if (!value.fields || !value.fields.timestamp) {
      return soFar;
    }
    const {timestamp} = value.fields;
    // Order messages by "virtual" nano-seconds
    if (timestamp === soFar.lastTimestamp) {
      soFar.currentNanoSecond += 1;
    } else {
      soFar.currentNanoSecond = parseInt(timestamp.slice(-6), 10);
    }
    value.timestamp = `${timestamp.slice(0, -6)}${padStart(soFar.currentNanoSecond.toString(), 6, '0')}`;
    soFar.uniquePoints.push(value);
    soFar.lastTimestamp = timestamp;
    return soFar;
  }, initialValue);

  return uniquePoints;
};

export default withUniqueTimestamps;
