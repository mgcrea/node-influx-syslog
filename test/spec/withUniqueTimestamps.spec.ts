import withUniqueTimestamps from 'src/utils/withUniqueTimestamps';
import {IPoint} from 'influx';

describe('withUniqueTimestamps', () => {
  it('should properly export a class', () => {
    expect(withUniqueTimestamps).toBeDefined();
    expect(typeof withUniqueTimestamps).toBe('function');
  });
  it('should properly order points by nanoseconds', () => {
    const timestamp = `${new Date(1562361784183).getTime()}000000`;
    const points: IPoint[] = [{fields: {timestamp}}, {fields: {timestamp}}, {fields: {timestamp}}];
    expect(withUniqueTimestamps(points)).toMatchSnapshot();
  });
  it('should properly order points by nanoseconds 2', () => {
    const timestamp = `${new Date(1562361784183).getTime()}000123`;
    const points: IPoint[] = [{fields: {timestamp}}, {fields: {timestamp}}, {fields: {timestamp}}];
    expect(withUniqueTimestamps(points)).toMatchSnapshot();
  });
  it('should properly order points by nanoseconds 3', () => {
    const timestampA = `${new Date(1562361784183).getTime()}000123`;
    const timestampB = `${new Date(1562361784183).getTime()}000456`;
    const timestampC = `${new Date(1562361784185).getTime()}000789`;
    const points: IPoint[] = [
      {fields: {timestamp: timestampA}},
      {fields: {timestamp: timestampA}},
      {fields: {timestamp: timestampB}},
      {fields: {timestamp: timestampC}},
      {fields: {timestamp: timestampC}},
      {fields: {timestamp: timestampC}}
    ];
    expect(withUniqueTimestamps(points)).toMatchSnapshot();
  });
});
