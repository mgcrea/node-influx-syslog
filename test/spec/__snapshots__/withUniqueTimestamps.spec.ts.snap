// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withUniqueTimestamps should properly order points by nanoseconds 1`] = `
Array [
  Object {
    "fields": Object {
      "timestamp": "1562361784183000000",
    },
    "timestamp": "1562361784183000000",
  },
  Object {
    "fields": Object {
      "timestamp": "1562361784183000000",
    },
    "timestamp": "1562361784183000001",
  },
  Object {
    "fields": Object {
      "timestamp": "1562361784183000000",
    },
    "timestamp": "1562361784183000002",
  },
]
`;

exports[`withUniqueTimestamps should properly order points by nanoseconds 2 1`] = `
Array [
  Object {
    "fields": Object {
      "timestamp": "1562361784183000123",
    },
    "timestamp": "1562361784183000123",
  },
  Object {
    "fields": Object {
      "timestamp": "1562361784183000123",
    },
    "timestamp": "1562361784183000124",
  },
  Object {
    "fields": Object {
      "timestamp": "1562361784183000123",
    },
    "timestamp": "1562361784183000125",
  },
]
`;

exports[`withUniqueTimestamps should properly order points by nanoseconds 3 1`] = `
Array [
  Object {
    "fields": Object {
      "timestamp": "1562361784183000123",
    },
    "timestamp": "1562361784183000123",
  },
  Object {
    "fields": Object {
      "timestamp": "1562361784183000123",
    },
    "timestamp": "1562361784183000124",
  },
  Object {
    "fields": Object {
      "timestamp": "1562361784183000456",
    },
    "timestamp": "1562361784183000456",
  },
  Object {
    "fields": Object {
      "timestamp": "1562361784185000789",
    },
    "timestamp": "1562361784185000789",
  },
  Object {
    "fields": Object {
      "timestamp": "1562361784185000789",
    },
    "timestamp": "1562361784185000790",
  },
  Object {
    "fields": Object {
      "timestamp": "1562361784185000789",
    },
    "timestamp": "1562361784185000791",
  },
]
`;
